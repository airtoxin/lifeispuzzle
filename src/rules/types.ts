import { Context, Bool } from "z3-solver";
import { BoardVariable } from "../states";

// ルールの基底インターフェース

export interface Rule {
  id: string;
  name: string;
  description: string;
  getConstraints<T extends string>(
    boardVar: BoardVariable<T>,
    ctx: Context<T>,
  ): Bool<T>[];
}
