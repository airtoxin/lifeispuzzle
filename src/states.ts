import { Arith, Bool, Context, init, Model } from "z3-solver";

// シリアライズ可能な盤面状態（基底型）
export interface BoardState {
  size: number;
  cells: number[][]; // セル値
  horizontalEdges: number[][]; // 水平線 (size+1 × size)
  verticalEdges: number[][]; // 垂直線 (size × size+1)
}

// Z3変数を使った盤面状態（BoardStateから自動導出）
export type BoardVariable<T extends string> = {
  [K in keyof BoardState]: K extends
    | "cells"
    | "horizontalEdges"
    | "verticalEdges"
    ? Arith<T>[][]
    : BoardState[K];
};

export type DeepReadonly<T> = keyof T extends never
  ? T
  : { readonly [K in keyof T]: DeepReadonly<T[K]> };

// 型変換ユーティリティ関数
export function createBoardVariable<T extends string>(
  boardState: BoardState,
  ctx: Context<T>,
): BoardVariable<T> {
  return {
    size: boardState.size,
    cells: boardState.cells.map((row, rowIndex) =>
      row.map((_, colIndex) => ctx.Int.const(`c-${rowIndex}-${colIndex}`)),
    ),
    horizontalEdges: boardState.horizontalEdges.map((row, rowIndex) =>
      row.map((_, colIndex) => ctx.Int.const(`he-${rowIndex}-${colIndex}`)),
    ),
    verticalEdges: boardState.verticalEdges.map((row, rowIndex) =>
      row.map((_, colIndex) => ctx.Int.const(`ve-${rowIndex}-${colIndex}`)),
    ),
  };
}

if (import.meta.vitest) {
  const { it, expect, describe } = import.meta.vitest;

  describe("createBoardVariable", () => {
    it("should have proper structure", async () => {
      const z3 = await import("z3-solver");
      const { Context } = await z3.init();
      const ctx = Context("test");

      const boardState: BoardState = {
        size: 2,
        cells: [
          [0, 0],
          [0, 0],
        ],
        horizontalEdges: [
          [0, 0],
          [0, 0],
          [0, 0],
        ],
        verticalEdges: [
          [0, 0, 0],
          [0, 0, 0],
        ],
      };

      const boardVar = createBoardVariable(boardState, ctx);

      expect(boardVar.size).toBe(2);
      expect(boardVar.cells.length).toBe(2);
      expect(boardVar.cells[0].length).toBe(2);
    });
  });
}

export function boardVariableToState<T extends string>(
  boardVar: BoardVariable<T>,
  model: Model<T>,
): BoardState {
  return {
    size: boardVar.size,
    cells: boardVar.cells.map((row) =>
      row.map((cellVar) => {
        const value = model.eval(cellVar);
        return parseInt(value.toString());
      }),
    ),
    horizontalEdges: boardVar.horizontalEdges.map((row) =>
      row.map((edgeVar) => {
        const value = model.eval(edgeVar);
        return parseInt(value.toString());
      }),
    ),
    verticalEdges: boardVar.verticalEdges.map((row) =>
      row.map((edgeVar) => {
        const value = model.eval(edgeVar);
        return parseInt(value.toString());
      }),
    ),
  };
}

if (import.meta.vitest) {
  const { it, expect, describe } = import.meta.vitest;

  describe("boardVariableToState", () => {
    it("should have proper structure", () => {
      const boardState: BoardState = {
        size: 2,
        cells: [
          [1, 2],
          [3, 4],
        ],
        horizontalEdges: [
          [0, 0],
          [0, 0],
          [0, 0],
        ],
        verticalEdges: [
          [0, 0, 0],
          [0, 0, 0],
        ],
      };

      expect(boardState.size).toBe(2);
      expect(boardState.cells.length).toBe(2);
      expect(boardState.cells[0].length).toBe(2);
    });
  });
}

if (import.meta.vitest) {
  const { it, expect, describe } = import.meta.vitest;

  describe("BoardState", () => {
    it("should be serializable to JSON", () => {
      const boardState: BoardState = {
        size: 2,
        cells: [
          [1, 2],
          [3, 4],
        ],
        horizontalEdges: [
          [0, 0],
          [0, 0],
          [0, 0],
        ],
        verticalEdges: [
          [0, 0, 0],
          [0, 0, 0],
        ],
      };

      const json = JSON.stringify(boardState);
      const parsed = JSON.parse(json) as BoardState;

      expect(parsed.size).toBe(2);
      expect(parsed.cells).toEqual([
        [1, 2],
        [3, 4],
      ]);
    });
  });
}
