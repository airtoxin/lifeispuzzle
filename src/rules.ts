import { BoardState, BoardVariable } from "./states.js";
import { Context, Bool } from "z3-solver";

// ルールの基底インターフェース
export interface Rule {
  id: string;
  name: string;
  description: string;
  getConstraints(boardVar: BoardVariable, ctx: Context<any>): Bool<any>[];
}

export const NumberFillRule: Rule = {
  id: "number-fill-rule",
  name: "数字で充填されている",
  description: "盤面の全てのマスが数字で埋められていること",
  getConstraints(boardVar, ctx) {
    return boardVar.cells.flat().map((v) => v.ge(1));
  },
};

export const RowUniquenessRule: Rule = {
  id: "row-uniqueness-rule",
  name: "行内数字一意性",
  description: "各行には同じ数字が複数現れない",
  getConstraints(boardVar, ctx) {
    return boardVar.cells.map((row) => ctx.Distinct(...row));
  },
};

export const ColumnUniquenessRule: Rule = {
  id: "column-uniqueness-rule",
  name: "列内数字一意性",
  description: "各列には同じ数字が複数現れない",
  getConstraints(boardVar, ctx) {
    return Array.from({ length: boardVar.size }, (_, colIndex) =>
      ctx.Distinct(...boardVar.cells.map((row) => row[colIndex]!)),
    );
  },
};

export const ColumnSortRule: Rule = {
  id: "column-sort-rule",
  name: "列ソート制約",
  description: "各列の数値が昇順または降順でソート済み",
  getConstraints(boardVar, ctx) {
    return Array.from({ length: boardVar.size }, (_, colIndex) => {
      const column = boardVar.cells.map((row) => row[colIndex]!);

      // 昇順制約: c[0] ≤ c[1] ≤ c[2] ≤ ...
      const ascending = column.slice(1).map((curr, i) => column[i].le(curr));

      // 降順制約: c[0] ≥ c[1] ≥ c[2] ≥ ...
      const descending = column.slice(1).map((curr, i) => column[i].ge(curr));

      // 昇順または降順
      return ctx.Or(ctx.And(...ascending), ctx.And(...descending));
    });
  },
};

export const RowSortRule: Rule = {
  id: "row-sort-rule",
  name: "行ソート制約",
  description: "各行の数値が昇順または降順でソート済み",
  getConstraints(boardVar, ctx) {
    return boardVar.cells.map((row) => {
      // 昇順制約: r[0] ≤ r[1] ≤ r[2] ≤ ...
      const ascending = row.slice(1).map((curr, i) => row[i].le(curr));

      // 降順制約: r[0] ≥ r[1] ≥ r[2] ≥ ...
      const descending = row.slice(1).map((curr, i) => row[i].ge(curr));

      // 昇順または降順
      return ctx.Or(ctx.And(...ascending), ctx.And(...descending));
    });
  },
};

export const MagicSquareRule: Rule = {
  id: "magic-square-rule",
  name: "魔法陣制約",
  description: "タテ・ヨコ・ナナメの合計が全て等しい",
  getConstraints(boardVar, ctx) {
    const constraints: any[] = [];
    // すべての数字が異なる
    constraints.push(ctx.Distinct(...boardVar.cells.flat()));
    constraints.push(
      ...boardVar.cells
        .flat()
        .map((v) => v.ge(1).and(v.le(boardVar.size * boardVar.size))),
    );

    // 最初の行の合計を基準とする
    const firstRowSum = boardVar.cells[0]
      .slice(1)
      .reduce((sum, cell) => sum.add(cell), boardVar.cells[0][0]);

    // 各行の合計が基準と等しい
    boardVar.cells.slice(1).forEach((row) => {
      const rowSum = row.slice(1).reduce((sum, cell) => sum.add(cell), row[0]);
      constraints.push(rowSum.eq(firstRowSum));
    });

    // 各列の合計が基準と等しい
    Array.from({ length: boardVar.size }, (_, colIndex) => {
      const column = boardVar.cells.map((row) => row[colIndex]);
      const columnSum = column
        .slice(1)
        .reduce((sum, cell) => sum.add(cell), column[0]);
      constraints.push(columnSum.eq(firstRowSum));
    });

    // 左上から右下への対角線の合計が基準と等しい
    const mainDiagonal = Array.from(
      { length: boardVar.size },
      (_, i) => boardVar.cells[i][i],
    );
    const mainDiagonalSum = mainDiagonal
      .slice(1)
      .reduce((sum, cell) => sum.add(cell), mainDiagonal[0]);
    constraints.push(mainDiagonalSum.eq(firstRowSum));

    // 右上から左下への対角線の合計が基準と等しい
    const antiDiagonal = Array.from(
      { length: boardVar.size },
      (_, i) => boardVar.cells[i][boardVar.size - 1 - i],
    );
    const antiDiagonalSum = antiDiagonal
      .slice(1)
      .reduce((sum, cell) => sum.add(cell), antiDiagonal[0]);
    constraints.push(antiDiagonalSum.eq(firstRowSum));

    return constraints;
  },
};

// 与えられた値の制約ルール
export function createGivenValuesRule(initialState: BoardState): Rule {
  return {
    id: "given-values-rule",
    name: "与えられた値制約",
    description: "初期値として与えられたセルの値を固定する",
    getConstraints(boardVar, ctx) {
      const constraints: any[] = [];

      for (let row = 0; row < initialState.size; row++) {
        for (let col = 0; col < initialState.size; col++) {
          const givenValue = initialState.cells[row][col];
          if (givenValue !== 0) {
            constraints.push(boardVar.cells[row][col].eq(givenValue));
          }
        }
      }

      return constraints;
    },
  };
}

// In-source tests
if (import.meta.vitest) {
  const { it, expect, describe } = import.meta.vitest;

  describe("Rules", () => {
    it("NumberFillRule should have correct properties", () => {
      expect(NumberFillRule.id).toBe("number-fill-rule");
      expect(NumberFillRule.name).toBe("数字で充填されている");
      expect(typeof NumberFillRule.getConstraints).toBe("function");
    });

    it("createGivenValuesRule should create rule with given values", () => {
      const initialBoardState: BoardState = {
        size: 2,
        cells: [
          [1, 0],
          [0, 4],
        ], // 1と4を固定
        horizontalEdges: [
          [0, 0],
          [0, 0],
          [0, 0],
        ],
        verticalEdges: [
          [0, 0, 0],
          [0, 0, 0],
        ],
      };

      const rule = createGivenValuesRule(initialBoardState);
      expect(rule.id).toBe("given-values-rule");
      expect(rule.name).toBe("与えられた値制約");
      expect(typeof rule.getConstraints).toBe("function");
    });

    it("MagicSquareRule should have expected constraint count structure", async () => {
      // この部分は実際のZ3初期化なしでの構造テスト
      expect(MagicSquareRule.id).toBe("magic-square-rule");
      expect(MagicSquareRule.description).toContain("タテ・ヨコ・ナナメ");
    });
  });
}
