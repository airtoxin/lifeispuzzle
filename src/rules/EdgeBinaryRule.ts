import { Context, Bool } from "z3-solver";
import { Rule } from "./types";
import { BoardState, BoardVariable } from "../states";
import { createGivenEdgesRule } from "./helpers";

// スリザーリンク用ルール

export const EdgeBinaryRule: Rule = {
  id: "edge-binary-rule",
  name: "エッジバイナリ制約",
  description: "エッジが0または1の値のみを持つ",
  getConstraints<T extends string>(
    boardVar: BoardVariable<T>,
    ctx: Context<T>,
  ) {
    const constraints: Bool<T>[] = [];

    // 水平エッジの制約
    boardVar.horizontalEdges.forEach((row) => {
      row.forEach((edge) => {
        constraints.push(edge.ge(0).and(edge.le(1)));
      });
    });

    // 垂直エッジの制約
    boardVar.verticalEdges.forEach((row) => {
      row.forEach((edge) => {
        constraints.push(edge.ge(0).and(edge.le(1)));
      });
    });

    return constraints;
  },
};
if (import.meta.vitest) {
  const { it, expect, describe } = import.meta.vitest;

  describe("EdgeBinaryRule", () => {
    it("should be satisfied with valid edge values (0 and 1)", async () => {
      const z3 = await import("z3-solver");
      const { Context } = await z3.init();
      const ctx = Context("test");
      const { createBoardVariable } = await import("../states.js");

      // 有効なエッジ値（0と1のみ）を持つ盤面
      const validBoardState: BoardState = {
        size: 2,
        cells: [
          [0, 0],
          [0, 0],
        ],
        horizontalEdges: [
          [1, 0],
          [0, 1],
          [1, 0],
        ],
        verticalEdges: [
          [0, 1, 0],
          [1, 0, 1],
        ],
      };

      const boardVar = createBoardVariable(validBoardState, ctx);
      const edgeBinaryConstraints = EdgeBinaryRule.getConstraints(
        boardVar,
        ctx,
      );
      const givenEdgeConstraints = createGivenEdgesRule(
        validBoardState,
      ).getConstraints(boardVar, ctx);

      const solver = new ctx.Solver();
      [...edgeBinaryConstraints, ...givenEdgeConstraints].forEach(
        (constraint) => solver.add(constraint),
      );

      const result = await solver.check();
      expect(result).toBe("sat");
    });

    it("should be violated with invalid edge values (outside 0-1 range)", async () => {
      const z3 = await import("z3-solver");
      const { Context } = await z3.init();
      const ctx = Context("test");
      const { createBoardVariable } = await import("../states.js");

      // 無効なエッジ値を含む盤面
      const invalidBoardState: BoardState = {
        size: 2,
        cells: [
          [0, 0],
          [0, 0],
        ],
        horizontalEdges: [
          [2, 0], // 2は無効（0-1の範囲外）
          [0, 0],
          [0, 0],
        ],
        verticalEdges: [
          [0, 0, 0],
          [0, 0, 0],
        ],
      };

      const boardVar = createBoardVariable(invalidBoardState, ctx);
      const edgeBinaryConstraints = EdgeBinaryRule.getConstraints(
        boardVar,
        ctx,
      );
      const givenEdgeConstraints = createGivenEdgesRule(
        invalidBoardState,
      ).getConstraints(boardVar, ctx);

      const solver = new ctx.Solver();
      [...edgeBinaryConstraints, ...givenEdgeConstraints].forEach(
        (constraint) => solver.add(constraint),
      );

      const result = await solver.check();
      expect(result).toBe("unsat");
    });
  });
}
