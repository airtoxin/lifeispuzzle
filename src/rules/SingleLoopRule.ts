import { Context, Bool, Arith } from "z3-solver";
import { isDegree2, getNeighbors, hasEdgeBetween } from "./AllEdgesInLoopRule";
import { Rule } from "./types";
import { BoardVariable } from "../states";
import { createGivenEdgesRule } from "./helpers";
import { VertexDegreeRule } from "./VertexDegreeRule";

export const SingleLoopRule: Rule = {
  id: "single-loop-rule",
  name: "単一ループ制約（到達可能性ベース）",
  description: "距離変数を使用して単一のループのみを許可",
  getConstraints<T extends string>(
    boardVar: BoardVariable<T>,
    ctx: Context<T>,
  ) {
    const constraints: Bool<T>[] = [];
    const size = boardVar.size;

    // 距離変数ベースの単一ループ制約
    // 各頂点の距離変数（ルートからの距離、-1はループに含まれない頂点）
    const dist: Arith<T>[][] = [];
    for (let p = 0; p <= size; p++) {
      dist[p] = [];
      for (let q = 0; q <= size; q++) {
        dist[p][q] = ctx.Int.const(`dist_${p}_${q}`);
        // 距離は-1以上、最大でもループサイズ以下
        constraints.push(dist[p][q].ge(-1));
        constraints.push(dist[p][q].le((size + 1) * 2)); // 最大ループサイズ
      }
    }

    // 制約1: ループに含まれない頂点（次数が2でない）の距離は-1
    for (let p = 0; p <= size; p++) {
      for (let q = 0; q <= size; q++) {
        const deg2 = isDegree2(boardVar, ctx, p, q);
        constraints.push(ctx.Implies(ctx.Not(deg2), dist[p][q].eq(-1)));
      }
    }

    // 制約2: ルートの一意性（距離0の頂点は最大1個）
    let rootCount: Arith<T> = ctx.Int.val(0);
    for (let p = 0; p <= size; p++) {
      for (let q = 0; q <= size; q++) {
        const deg2 = isDegree2(boardVar, ctx, p, q);
        const isRoot = ctx.And(deg2, dist[p][q].eq(0));
        rootCount = rootCount.add(
          ctx.If(isRoot, ctx.Int.val(1), ctx.Int.val(0)),
        );
      }
    }

    // 次数2の頂点が存在するかチェック
    let hasDeg2Vertex: Bool<T> = ctx.Bool.val(false);
    for (let p = 0; p <= size; p++) {
      for (let q = 0; q <= size; q++) {
        const deg2 = isDegree2(boardVar, ctx, p, q);
        hasDeg2Vertex = ctx.Or(hasDeg2Vertex, deg2);
      }
    }

    // ループが存在する場合はルートが1個、存在しない場合はルートが0個
    constraints.push(ctx.Implies(hasDeg2Vertex, rootCount.eq(1)));
    constraints.push(ctx.Implies(ctx.Not(hasDeg2Vertex), rootCount.eq(0)));

    // 制約3: 親子関係制約（ループを考慮）
    // 次数2の各頂点について、隣接する次数2の頂点の中でちょうど1つが親（距離が1小さい）または子（距離が1大きい）である
    for (let p = 0; p <= size; p++) {
      for (let q = 0; q <= size; q++) {
        const deg2 = isDegree2(boardVar, ctx, p, q);

        // 隣接する次数2の頂点の中で、親または子を探す
        const neighbors = getNeighbors(boardVar, p, q);
        const hasParentOrChild: Bool<T>[] = [];

        for (const neighbor of neighbors) {
          const [np, nq] = neighbor;
          if (np >= 0 && np <= size && nq >= 0 && nq <= size) {
            const hasEdge = hasEdgeBetween(boardVar, ctx, p, q, np, nq);
            const neighborDeg2 = isDegree2(boardVar, ctx, np, nq);

            // 親の場合: 隣接頂点の距離 = 自分の距離 - 1
            const isParent = dist[np][nq].eq(dist[p][q].sub(1));
            // 子の場合: 隣接頂点の距離 = 自分の距離 + 1
            const isChild = dist[np][nq].eq(dist[p][q].add(1));
            // ループの特殊ケース: ルート（距離0）と最大距離の頂点の関係
            const isLoopConnection = ctx.And(
              ctx.Or(
                ctx.And(dist[p][q].eq(0), dist[np][nq].gt(0)),
                ctx.And(dist[np][nq].eq(0), dist[p][q].gt(0)),
              ),
            );

            // エッジが存在し、隣接頂点が次数2で、親子関係またはループ接続がある
            hasParentOrChild.push(
              ctx.And(
                hasEdge,
                neighborDeg2,
                ctx.Or(isParent, isChild, isLoopConnection),
              ),
            );
          }
        }

        if (hasParentOrChild.length > 0) {
          // 次数2の頂点は必ず隣接する次数2の頂点と適切な距離関係を持つ
          constraints.push(ctx.Implies(deg2, ctx.Or(...hasParentOrChild)));
        }
      }
    }

    // シンプルな制約: ループが存在する場合、最低4エッジ必要
    let totalEdges: Arith<T> = ctx.Int.val(0);

    // 水平エッジを数える
    for (let p = 0; p <= size; p++) {
      for (let q = 0; q < size; q++) {
        totalEdges = totalEdges.add(boardVar.horizontalEdges[p][q]);
      }
    }

    // 垂直エッジを数える
    for (let p = 0; p < size; p++) {
      for (let q = 0; q <= size; q++) {
        totalEdges = totalEdges.add(boardVar.verticalEdges[p][q]);
      }
    }

    // ループが存在する場合、最低4エッジ必要
    constraints.push(ctx.Implies(hasDeg2Vertex, totalEdges.ge(4)));

    return constraints;
  },
};
if (import.meta.vitest) {
  const { it, expect, describe } = import.meta.vitest;

  describe("SingleLoopRule (Reachability-based)", () => {
    it("should be satisfied with no edges", async () => {
      const z3 = await import("z3-solver");
      const { Context } = await z3.init();
      const ctx = Context("test");
      const { createBoardVariable } = await import("../states.js");

      const boardVar = createBoardVariable(
        {
          size: 2,
          cells: [
            [0, 0],
            [0, 0],
          ],
          horizontalEdges: [
            [0, 0],
            [0, 0],
            [0, 0],
          ],
          verticalEdges: [
            [0, 0, 0],
            [0, 0, 0],
          ],
        },
        ctx,
      );

      const singleLoopConstraints = SingleLoopRule.getConstraints(
        boardVar,
        ctx,
      );

      const solver = new ctx.Solver();
      singleLoopConstraints.forEach((constraint) => solver.add(constraint));

      // 全エッジを0に固定（エッジなし）
      boardVar.horizontalEdges.flat().forEach((edge) => solver.add(edge.eq(0)));
      boardVar.verticalEdges.flat().forEach((edge) => solver.add(edge.eq(0)));

      const result = await solver.check();
      expect(result).toBe("sat");
    });

    it("should be satisfied with a valid single loop", async () => {
      const z3 = await import("z3-solver");
      const { Context } = await z3.init();
      const ctx = Context("test");
      const { createBoardVariable } = await import("../states.js");

      const boardVar = createBoardVariable(
        {
          size: 2,
          cells: [
            [0, 0],
            [0, 0],
          ],
          horizontalEdges: [
            [1, 1], // 上の行
            [0, 0], // 真ん中の行
            [1, 1], // 下の行
          ],
          verticalEdges: [
            [1, 0, 1], // 左と右の境界
            [1, 0, 1],
          ],
        },
        ctx,
      );

      const singleLoopConstraints = SingleLoopRule.getConstraints(
        boardVar,
        ctx,
      );
      const vertexDegreeConstraints = VertexDegreeRule.getConstraints(
        boardVar,
        ctx,
      );
      const givenEdgeConstraints = createGivenEdgesRule({
        size: 2,
        cells: [
          [0, 0],
          [0, 0],
        ],
        horizontalEdges: [
          [1, 1],
          [0, 0],
          [1, 1],
        ],
        verticalEdges: [
          [1, 0, 1],
          [1, 0, 1],
        ],
      }).getConstraints(boardVar, ctx);

      const solver = new ctx.Solver();
      [
        ...singleLoopConstraints,
        ...vertexDegreeConstraints,
        ...givenEdgeConstraints,
      ].forEach((constraint) => solver.add(constraint));

      const result = await solver.check();
      expect(result).toBe("sat");
    });

    it("should be violated with multiple disconnected loops", async () => {
      const z3 = await import("z3-solver");
      const { Context } = await z3.init();
      const ctx = Context("test");
      const { createBoardVariable } = await import("../states.js");

      const boardVar = createBoardVariable(
        {
          size: 3,
          cells: [
            [0, 0, 0],
            [0, 0, 0],
            [0, 0, 0],
          ],
          horizontalEdges: [
            [1, 1, 0],
            [0, 0, 0],
            [0, 0, 1],
            [0, 0, 1],
          ],
          verticalEdges: [
            [1, 1, 0, 0],
            [1, 1, 0, 0],
            [0, 0, 1, 1],
          ],
        },
        ctx,
      );

      const singleLoopConstraints = SingleLoopRule.getConstraints(
        boardVar,
        ctx,
      );
      const vertexDegreeConstraints = VertexDegreeRule.getConstraints(
        boardVar,
        ctx,
      );

      // 2つの独立した小さなループを作る設定
      const givenEdgeConstraints = createGivenEdgesRule({
        size: 3,
        cells: [
          [0, 0, 0],
          [0, 0, 0],
          [0, 0, 0],
        ],
        horizontalEdges: [
          [1, 1, 0], // 左上のループ
          [0, 0, 0],
          [0, 0, 1], // 右下のループ
          [0, 0, 1],
        ],
        verticalEdges: [
          [1, 1, 0, 0],
          [1, 1, 0, 0],
          [0, 0, 1, 1],
        ],
      }).getConstraints(boardVar, ctx);

      const solver = new ctx.Solver();
      [
        ...singleLoopConstraints,
        ...vertexDegreeConstraints,
        ...givenEdgeConstraints,
      ].forEach((constraint) => solver.add(constraint));

      const result = await solver.check();
      expect(result).toBe("unsat");
    });
  });
}
