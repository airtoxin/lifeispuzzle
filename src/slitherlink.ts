import { Context, Bool } from "z3-solver";
import { Rule } from "./rules/types";
import { BoardState, BoardVariable, createBoardVariable, boardVariableToState } from "./states";

// スリザーリンクのセルの数字制約ルール
export const SlitherlinkNumberConstraintRule: Rule = {
  id: "slitherlink-number-constraint-rule",
  name: "スリザーリンクの数字制約",
  description: "各セルの数字はその周りのエッジの数を示す",
  getConstraints<T extends string>(
    boardVar: BoardVariable<T>,
    ctx: Context<T>,
  ) {
    const constraints: Bool<T>[] = [];
    const size = boardVar.size;

    for (let row = 0; row < size; row++) {
      for (let col = 0; col < size; col++) {
        const cellValue = boardVar.cells[row][col];
        
        // セルの周りのエッジを収集
        const edges: Array<typeof boardVar.horizontalEdges[0][0]> = [];
        
        // 上のエッジ (水平)
        edges.push(boardVar.horizontalEdges[row][col]);
        // 下のエッジ (水平)
        edges.push(boardVar.horizontalEdges[row + 1][col]);
        // 左のエッジ (垂直)
        edges.push(boardVar.verticalEdges[row][col]);
        // 右のエッジ (垂直)
        edges.push(boardVar.verticalEdges[row][col + 1]);

        // セルの値が0でない場合、エッジの合計がセルの値と等しい
        const edgeSum = edges.reduce((sum, edge) => sum.add(edge));
        
        // セルの値が0の場合は制約なし、0でない場合はエッジの合計と等しい
        constraints.push(
          cellValue.eq(0).implies(ctx.Bool.val(true)).and(
            cellValue.neq(0).implies(edgeSum.eq(cellValue))
          )
        );
      }
    }

    return constraints;
  },
};

// 頂点次数制約ルール（既存のものを再エクスポート）
import { VertexDegreeRule } from "./rules/VertexDegreeRule";
export { VertexDegreeRule };

// エッジバイナリルール（既存のものを再エクスポート）
import { EdgeBinaryRule } from "./rules/EdgeBinaryRule";
export { EdgeBinaryRule };

// スリザーリンク単一ループルール（既存のものを再エクスポート）
import { SingleLoopRule } from "./rules/SingleLoopRule";
export { SingleLoopRule };

// スリザーリンクの全ルール
export const SlitherlinkRules = [
  EdgeBinaryRule,
  SlitherlinkNumberConstraintRule,
  VertexDegreeRule,
  SingleLoopRule,
];

// スリザーリンクソルバー
export interface SlitherlinkPuzzle {
  size: number;
  clues: (number | null)[][]; // nullは空セル
}

export function createSlitherlinkBoardState(puzzle: SlitherlinkPuzzle): BoardState {
  const size = puzzle.size;
  return {
    size,
    cells: puzzle.clues.map(row => 
      row.map(clue => clue === null ? 0 : clue)
    ),
    horizontalEdges: Array(size + 1).fill(null).map(() => Array(size).fill(0)),
    verticalEdges: Array(size).fill(null).map(() => Array(size + 1).fill(0)),
  };
}

export async function solveSlitherlink(puzzle: SlitherlinkPuzzle): Promise<BoardState | null> {
  const z3 = await import("z3-solver");
  const { Context } = await z3.init();
  const ctx = Context("slitherlink");

  const boardState = createSlitherlinkBoardState(puzzle);
  const boardVar = createBoardVariable(boardState, ctx);

  const solver = new ctx.Solver();

  // 与えられたクリューの制約を追加
  const { createGivenValuesRule } = await import("./rules/helpers");
  const givenValuesRule = createGivenValuesRule(boardState);
  const givenConstraints = givenValuesRule.getConstraints(boardVar, ctx);
  givenConstraints.forEach(constraint => solver.add(constraint));

  // スリザーリンクのルールを追加
  SlitherlinkRules.forEach(rule => {
    const constraints = rule.getConstraints(boardVar, ctx);
    constraints.forEach(constraint => solver.add(constraint));
  });

  const result = await solver.check();
  if (result === "sat") {
    const model = solver.model();
    return boardVariableToState(boardVar, model);
  } else {
    return null;
  }
}

// 結果をビジュアライズする関数
export function visualizeSlitherlinkSolution(solution: BoardState): string {
  const size = solution.size;
  let result = "";

  for (let row = 0; row <= size; row++) {
    // 頂点と水平エッジの行
    let line = "";
    for (let col = 0; col <= size; col++) {
      // 頂点
      line += "+";
      
      // 水平エッジ（最後の列でない場合）
      if (col < size && row < size + 1) {
        const hasEdge = solution.horizontalEdges[row][col] === 1;
        line += hasEdge ? "---" : "   ";
      }
    }
    result += line + "\n";

    // セルと垂直エッジの行（最後の行でない場合）
    if (row < size) {
      line = "";
      for (let col = 0; col <= size; col++) {
        // 垂直エッジ
        const hasEdge = solution.verticalEdges[row][col] === 1;
        line += hasEdge ? "|" : " ";
        
        // セル（最後の列でない場合）
        if (col < size) {
          const cellValue = solution.cells[row][col];
          const cellDisplay = cellValue === 0 ? " " : cellValue.toString();
          line += ` ${cellDisplay} `;
        }
      }
      result += line + "\n";
    }
  }

  return result;
}

if (import.meta.vitest) {
  const { it, expect, describe } = import.meta.vitest;

  describe("SlitherlinkNumberConstraintRule", () => {
    it("should enforce number constraints correctly", async () => {
      const z3 = await import("z3-solver");
      const { Context } = await z3.init();
      const ctx = Context("test");

      // 2x2の例：数字3がある場合、周りの4つのエッジのうち3つが1である必要がある
      const boardState: BoardState = {
        size: 2,
        cells: [
          [3, 0],
          [0, 0],
        ],
        horizontalEdges: [
          [1, 0], // 上の行
          [1, 0], // 真ん中の行
          [1, 0], // 下の行
        ],
        verticalEdges: [
          [1, 0, 0], // 左の列、真ん中の列、右の列
          [0, 0, 0],
        ],
      };

      const boardVar = createBoardVariable(boardState, ctx);
      const constraints = SlitherlinkNumberConstraintRule.getConstraints(boardVar, ctx);
      
      const { createGivenValuesRule } = await import("./rules/helpers");
      const givenConstraints = createGivenValuesRule(boardState).getConstraints(boardVar, ctx);

      const solver = new ctx.Solver();
      [...constraints, ...givenConstraints].forEach(constraint => solver.add(constraint));

      const result = await solver.check();
      expect(result).toBe("sat");
    });

    it("should reject invalid number constraints", async () => {
      const z3 = await import("z3-solver");
      const { Context } = await z3.init();
      const ctx = Context("test");

      // 2x2の例：数字3があるが、すべてのエッジを0に固定（矛盾）
      const boardState: BoardState = {
        size: 2,
        cells: [
          [3, 0],
          [0, 0],
        ],
        horizontalEdges: [
          [0, 0], // 上の行 - すべて0に固定
          [0, 0], // 真ん中の行  
          [0, 0], // 下の行
        ],
        verticalEdges: [
          [0, 0, 0], // 左の列、真ん中の列、右の列 - すべて0に固定
          [0, 0, 0],
        ],
      };

      const boardVar = createBoardVariable(boardState, ctx);
      const constraints = SlitherlinkNumberConstraintRule.getConstraints(boardVar, ctx);
      
      const { createGivenValuesRule, createGivenEdgesRule } = await import("./rules/helpers");
      const givenConstraints = createGivenValuesRule(boardState).getConstraints(boardVar, ctx);
      const edgeConstraints = createGivenEdgesRule(boardState).getConstraints(boardVar, ctx);

      const solver = new ctx.Solver();
      [...constraints, ...givenConstraints, ...edgeConstraints].forEach(constraint => solver.add(constraint));

      const result = await solver.check();
      expect(result).toBe("unsat");
    });
  });


  describe("solveSlitherlink", () => {
    it("should solve the 3x3 puzzle correctly", async () => {
      const puzzle: SlitherlinkPuzzle = {
        size: 3,
        clues: [
          [null, 3, null],
          [3, null, 3],
          [null, 3, null],
        ],
      };

      const solution = await solveSlitherlink(puzzle);
      expect(solution).not.toBeNull();
      
      if (solution) {
        // 解が見つかった場合、ビジュアライズを確認
        const visualization = visualizeSlitherlinkSolution(solution);
        console.log("3x3 Slitherlink Solution:");
        console.log(visualization);
        
        // クリューが正しく解決されているか確認
        expect(solution.cells[0][1]).toBe(3); // クリュー: 3
        expect(solution.cells[1][0]).toBe(3); // クリュー: 3
        expect(solution.cells[1][2]).toBe(3); // クリュー: 3
        expect(solution.cells[2][1]).toBe(3); // クリュー: 3
      }
    });

    it("should visualize solution correctly", async () => {
      // 簡単な2x2の解をテスト
      const solution: BoardState = {
        size: 2,
        cells: [
          [2, 2],
          [2, 2],
        ],
        horizontalEdges: [
          [1, 1], // 上の行
          [0, 0], // 真ん中の行
          [1, 1], // 下の行
        ],
        verticalEdges: [
          [1, 0, 1], // 左の列、真ん中の列、右の列
          [1, 0, 1],
        ],
      };

      const visualization = visualizeSlitherlinkSolution(solution);
      expect(visualization).toMatchSnapshot();
    });
  });
}