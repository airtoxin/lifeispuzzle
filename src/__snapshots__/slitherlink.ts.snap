// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solveSlitherlink > should visualize solution correctly 1`] = `
"+---+---+
| 2 | 2 |
+   +   +
| 2 | 2 |
+---+---+
"
`;